src/msg/async/AsyncMessenger.cc

```diff
@@ -501,22 +501,28 @@ void AsyncMessenger::_finish_bind(const entity_addrvec_t& bind_addrs,
   if (get_myaddrs().front().get_port() == 0) {
     set_myaddrs(listen_addrs);
   }
-  entity_addrvec_t newaddrs = *my_addrs;
-  for (auto& a : newaddrs.v) {
-    a.set_nonce(nonce);
-    if (saved_public_addrs) {
-      // transplantate network layer addresses while keeping ports
-      // (as they can be figured out by msgr from the allowed range [1])
-      // unless they are explicitly specified (NATing both IP/port?)
-      //
-      // [1]: the low-level `Processor::bind` scans for free ports in
-      // a range controlled by ms_bind_port_min and ms_bind_port_max
-      const auto& public_addr =
-        saved_public_addrs->addr_of_type(a.get_type());
-      const auto public_port = public_addr.get_port();
-      const auto bound_port = a.get_port();
-      a.set_sockaddr(public_addr.get_sockaddr());
-      a.set_port(public_port == 0 ? bound_port : public_port);
+
+  entity_addrvec_t newaddrs;
+  if (saved_public_addrs) {
+    newaddrs = *saved_public_addrs;
+    for (auto& public_addr : newaddrs.v) {
+      public_addr.set_nonce(nonce);
+      if (public_addr.is_ip() && public_addr.get_port() == 0) {
+	// port is not explicitly set. This is fine as it can be figured
+	// out by msgr. For instance, the low-level `Processor::bind`
+	// scans for free ports in a range controlled by ms_bind_port_min
+	// and ms_bind_port_max.
+        for (const auto& a : my_addrs->v) {
+          if (public_addr.get_type() == a.get_type() && a.is_ip()) {
+             public_addr.set_port(a.get_port());
+          }
+        }
```

Comment:
If this is being respinned anyway, why not get rid of the loop as well?  IMO by explicitly making a provision for setting the port multiple times it just brings confusion -- one has to wonder about the significance of the last matching `a` winning (as opposed to the first) or whether `set_port` has some side effects that necessitate it being called multiple times...
```suggestion
        const auto& my_addr = my_addrs->addr_of_type(public_addr.get_type());
        if (my_addr.is_ip()) {
          public_addr.set_port(my_addr.get_port());
        }
```

Reply:
Behind our QAs run, there is a lot of testing in other projects too. That's the reason why I don't want to introduce any more changes.

Reply:
I'm not insisting on this change being made but want to point out that these points, while they seemingly make sense in the immediate term that covers the merge this PR, are flawed otherwise.  Hesitation with regards to making a change like that means a) that this code too fragile and b) that the contract is either underspecified or poorly understood.  It also highlights the lack of (any?) test coverage in Ceph -- a unit test that would instantiate `Messenger` instances with various combinations of bind and public addresses is needed at the very least.

This change, which you admit to be simple, could be performed by someone during refactoring without a second thought.  Even today, few people are aware of what the "global IP" feature is and certainly no one is going to repeat Santosh's manual verification in the future.

Reply:
> OK, so we agree the `addr_of_type()` thing is not a blocker and the change won't be conveyed within this PR.

Yeah, definitely not a blocker.  I just didn't realize that this entire PR was hanging on a manual test performed externally.